Generate `non-coherent cache` loads and `noalias` for CUDA

grypp opened the pull request:
IREE can recognize `readonly` data. If I'm not mistaken, this indicates the data is read-only and not aliased with data stored in the same region/kernel.

This PR makes advantage of this information to generates the `llvm.noalias` and `llvm.readonly` attributes to CUDA kernel function parameters. If an argument is not readonly, it still set `llvm.noalias` because the ranges used within the bindings are guaranteed not to alias.

This is crucial for two reasons. When the downstream compiler detects that it can better schedule the load instructions. It typically schedules them together that utilized bandwith better. Second, it generates 'non-coherent cache loads' (`nc` prefix to `ld.global` instruction in PTX). What PTX models say about it is as follows (this part gives room for experimentation):

```
The texture cache is larger, has higher bandwidth, and longer latency than the global memory cache.
For applications with sufficient parallelism to cover the longer latency, ld.global.nc should offer 
better performance than ld.global.
```

This PR first extends `hal.interface.binding.subspan` with `readonly` argument.
```
hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) flags(ReadOnly) : memref<16xf32>
```

In convert nvvm pass, it leverages `flags(ReadOnly)` to decide `llvm.readonly` attribute on the function argument. It also marks `llvm.noalias` every other argument. 

```
llvm.func @foo(%arg0: !llvm.ptr<f32> {llvm.align = 16 : i32, llvm.noalias, llvm.readonly})
```

benvanik reviewed (changes requested):
Nice - really great to see this getting plumbed through!

We have flags for this we can use instead by adding an optional flags attr to the op instead - `HAL_DescriptorFlagsAttr` has `HAL_DescriptorFlags_ReadOnly` and in the future we should add bits for things like non-temporal, non-sequential, etc. Can you please change the op to be defined as
```
   HAL_DescriptorTypeAttr:$descriptor_type,
   OptionalAttr<HAL_DescriptorFlagsAttr>:$descriptor_flags,
```
instead? You're already fetching the flags in MaterializeInterfaces and can pass that on directly.

RE aliasing: the ranges used within the bindings are guaranteed not to alias (and if they ever did we'd have to add a MayAlias flag (we have IREE_HAL_MEMORY_ACCESS_MAY_ALIAS in the runtime but it's not in the compiler yet). This makes it possible to put restrict/noalias on all mutable bindings as we guarantee that even if the physical storage ranges of two bindings may have overlap we'll never be producing code that relies on that overlap (reading byte N of binding 0 will never alias with a logical write of byte M on binding 1). All that to say I don't think you need to condition noalias on whether the binding is read-only or not - if we ever don't want noalias we'd add the `HAL_DescriptorFlags_MayAlias` flag and use that instead :)
The other flag we'll need at some point is for indicating whether atomics are enabled indicating that either multiple workgroups in the same dispatch or multiple dispatches are accessing the memory - today we don't and it's safe to assume all mutable memory is dispatch-local and by construction each workgroup is the sole writer of any particular memory location.

If you're on an adventure here there are some other things that can be plumbed through like `stream.alignment` and `stream.values` in the future - today I believe they get dropped on the floor :)

benvanik reviewed (changes requested):
Some minor style nits but otherwise LGTM! Thanks for doing this!

ThomasRaoux reviewed (approved):
Looks good to me but please given @benvanik another chance to review.

grypp committed:
Generate `non-coherent cache` loads and `noalias` for CUDA

IREE can recognize `readonly` data. If I'm not mistaken, this indicates the data is read-only and not aliased with data stored in the same region/kernel.

This PR makes advantage of this information to generates the 'llvm.noalias' and 'llvm.readonly' attributes to CUDA kernel function parameters.

This is crucial for two reasons. When the downstream compiler detects that it can better schedule the load instructions. It typically schedules them together that utilized bandwith better. Second, it generates 'non-coherent cache loads' ('nc' prefix to 'ld.global' instruction in PTX). What PTX models say about it is as follows (this part gives room for experimentation):

```
The texture cache is larger, has higher bandwidth, and longer latency than the global memory cache.
For applications with sufficient parallelism to cover the longer latency, ld.global.nc should
offer better performance than ld.global.
```

This PR first extends `hal.interface.binding.subspan` with `readonly` argument.
```
hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) {readonly} : memref<16xf32>
```

In convert nvvm pass, it leverages `readonly` to decide `llvm.noalias` and `llvm.readonly` attribute on the function argument.

```
llvm.func @foo(%arg0: !llvm.ptr<f32> {llvm.align = 16 : i32, llvm.noalias, llvm.readonly})
```

ScottTodd commented on compiler/src/iree/compiler/Codegen/LLVMGPU/ConvertToLLVM.cpp:
```diff
@@ -309,6 +309,20 @@ class ConvertIREEBindingSubspanOp : public ConvertToLLVMPattern {
+                           IREE::HAL::DescriptorFlags::ReadOnly)) {
+      // Setting the readonly attribute here will generate non-coherent cache
+      // loads.
+      llvmFuncOp.setArgAttr(llvmBufferArg.getArgNumber(),
```
I think we're hitting the limits of this current implementation while trying to enable `--iree-stream-resource-alias-mutable-bindings=true` by default (https://github.com/openxla/iree/pull/14658). That flag shows promising results (10-15% latency improvements) on other backends and is required for correctness on WebGPU, but aliasing the constant and mutable bindings gets the CUDA target off of this fast path of `readonly` + `noalias`.

For example, without the flag, we can have IR like this before `ConvertToNVVM`:
```mlir
    %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c4348864) flags(ReadOnly) : memref<64x300x300xf32, strided<[90000, 300, 1], offset: 1087216>>
    memref.assume_alignment %0, 64 : memref<64x300x300xf32, strided<[90000, 300, 1], offset: 1087216>>
    %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c27388864) : memref<1x64x302x302xf32, strided<[5837056, 91204, 302, 1], offset: 6847216>>
    memref.assume_alignment %1, 64 : memref<1x64x302x302xf32, strided<[5837056, 91204, 302, 1], offset: 6847216>>
```

and with the flag, the bindings are aliased (`binding(1)` to `binding(0)`):
```mlir
    %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c4348864) : memref<64x300x300xf32, strided<[90000, 300, 1], offset: 1087216>>
    memref.assume_alignment %0, 64 : memref<64x300x300xf32, strided<[90000, 300, 1], offset: 1087216>>
    %1 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c27388864) : memref<1x64x302x302xf32, strided<[5837056, 91204, 302, 1], offset: 6847216>>
    memref.assume_alignment %1, 64 : memref<1x64x302x302xf32, strided<[5837056, 91204, 302, 1], offset: 6847216>>
```

Then after `ConvertToNVVM`, the signature changes from
```mlir
  llvm.func @forward_dispatch_2(%arg0: !llvm.ptr {llvm.align = 16 : i32, llvm.noalias, llvm.readonly}, %arg1: !llvm.ptr {llvm.align = 16 : i32, llvm.noalias}) {
```

to
```mlir
llvm.func @forward_dispatch_2(%arg0: !llvm.ptr {llvm.align = 16 : i32, llvm.noalias}) {
```
and none of the pointer loads keep the `readonly` attribute

> For future work, is it possible to cluster read-only subspanOps in the same kernel argument?
>
> Another idea is to mark the load as read-only. But here I should convey the region information of each subspanOp to llvm dialect and work with the pointers there. This is not trivial.

Are either of these approaches viable?

grypp replied:
What is the motivation of enabling aliasing for Nvidia GPUs?

ScottTodd replied:
> What is the motivation of enabling aliasing for Nvidia GPUs?

My feeling is that since we set up the resource bindings in `stream` (prior to `hal`), we'd prefer to construct them in a target-agnostic way. We would then expect each target (CPU, Vulkan with some target triple, CUDA, etc.) to be able to lower to efficient code. As it stands now, we have some target/device configurations that appear to benefit from aliasing (according to our benchmarks) and some that don't.
